_posts/2020-05-07-tuning-a-complex-graph/tuning-a-complex-graph.Rmd

```diff
@@ -0,0 +1,335 @@
+
+```{r}
+graph1$param_set$values$branch_learner.selection = "ranger"
+graph1$param_set$values$classif.ranger.mtry = 200
+graph1$param_set$values$classif.ranger.num.trees = 1000
+
+graph1$train(task_train)
+msr("classif.acc")$score(graph1$predict(task_test)[[1L]])
+```
+
+Note that we also could have wrapped our graph in a `r ref("GraphLearner", "GraphLearner")` and proceeded to use this as a learner in an `r ref("AutoTuner", "AutoTuner")`.
+
+## Proxy
+
+Instead of using branches to split our graph with respect to the learner and preprocessing options, we can also use `PipeOpProxy`.
+`PipeOpProxy` accepts a single `content` parameter that can contain any other `PipeOp` or `Graph`.
+This is extremely flexible in the sense that we do not have to specify our options during construction.
+However, the parameters of the contained `PipeOp` or `Graph` are no longer directly contained in the `ParamSet` of the resulting graph.
+Therefore, when tuning the graph, we do have to make use of a `trafo` function.
+
+```{r}
+graph2 =
+  po("colapply", param_vals = list(
+    applicator = function(x) log(x, base = 10))
+  ) %>>%
+  po("scale") %>>%
+  po("proxy")
+```
+
+This graph now looks like the following:
+
+```{r, graph2}
+graph2$plot()
+```
+
+At first, this may look like a linear graph. However, as the `content` parameter of `PipeOpProxy` can be tuned and set to contain any other `PipeOp` or `Graph`, this will allow for a similar non-linear graph as when doing branching.
+
+```{r}
+graph2$param_set$ids()
+```
+
+We can tune the graph by using the same search space as before. However, here the `trafo` function is of central importance to actually set our options and parameters:
+
+```{r}
+tune_ps2 = tune_ps1$clone(deep = TRUE)
+```
+
+The `trafo` function does all the work, i.e., selecting either the `PCA`/`ICA`-`LDA` or `ranger` as the `proxy.content` as well as setting the parameters of the respective preprocessing `PipeOp`s and `Learner`s.
+
+```{r}
+proxy_options = list(
+  pca_ica_lda =
+    ppl("branch", graphs = list(pca = po("pca"), ica = po("ica"))) %>>%
+    lrn("classif.lda"),
+  ranger = lrn("classif.ranger")
+)
+```
+
+Note that above we made use of the `branch` `r ref("ppl", "ppl")` which is syntactic sugar for doing branching. Of course we also could have use another nested `PipeOpProxy` to specify the preprocessing options (`"pca"` vs. `"ica"`) within `proxy_options` if for some reason we do not want to do branching at all.
+The `trafo` function then selects one of the `proxy_options` and sets the respective parameters for the `PCA`, `ICA`, `LDA` and `ranger`.
+Here, the argument `x` is a list which will contain sampled/selected parameters from our `ParamSet` (in our case, `tune_ps2`).
+
+```{r}
+tune_ps2$trafo = function(x, param_set) {
+  values_out = list(proxy.content = proxy_options[[x$branch_learner.selection]])
+  if (x$branch_learner.selection == "pca_ica_lda") {
```

Comment:
maybe make a `proxy_content_ps` varaible here so you don't have to access `values_out$proxy.content$param_set` all the time below.

Reply:
Shorter function (though more obscure, you judge) would be
```r
proxy_content = proxy_options[[x$branch_learner.selection]]
paramids = intersect(proxy_content$param_set$ids(), names(x))
proxy_content$param_set$values[paramids] = x[paramids]
list(values_out = list(proxy.content = proxy_content)
```